File: samtools/bench3/qualhist.py
#!/usr/bin/env python3 

import seaborn as sns
import pysam,sys

bamName = sys.argv[1]
plotName = sys.argv[2]

def histQuals(bamInName):
    allquals=[]
    bamIn = pysam.AlignmentFile(bamInName)
    print("Reading BAM")
    for a in bamIn:
        allquals.append(a.mapping_quality)

    print("Plotting")
    return sns.histplot(allquals)
    
hist = histQuals(bamName)
hist.figure.savefig(plotName)

# time ./qualhist.py data/PC-HLA-EG4.bam histpy.png
# Reading BAM
# Plotting
# 1.22s user 0.42s system 66% cpu 2.446 total


# time qualhist.py /Users/james/projects/hla/capture/bam38/PC-HLA-EG4.bam histpy.png
# Reading BAM
# Plotting
# qualhist.py /Users/james/projects/hla/capture/bam38/PC-HLA-EG4.bam histpy.png  
# 18.19s user 0.93s system 98% cpu 19.359 total


# This implies:
# Warmup time 0.95 sec .
# Time to complete after warmup: 17.06 sec

File: samtools/bench3/qualhist.jl
#!/usr/bin/env julia

using XAM,Plots

bamName = ARGS[1]
plotName = ARGS[2]

function histQuals(bamInName)    
    allquals=[]
	print("Reading BAM...")
    bamIn = open(BAM.Reader,bamInName)
    for a in bamIn
        push!(allquals,BAM.mappingquality(a))
    end
	println("done.")
	#print(Int.(allquals[1:100])) # Int. bit is so output is shown as ints not as hex
	
    histogram(allquals)
end


hist = histQuals(bamName)
print("Saving fig to ",plotName,"...")
savefig(hist,plotName)
println("done.")


# time ./qualhist.jl data/PC-HLA-EG4.bam histjl.png
# Reading BAM...done.
# Saving fig to histjl.png...done.
# 
# 14.32s user 0.78s system 99% cpu 15.241 total


# time qualhist.jl /Users/james/projects/hla/capture/bam38/PC-HLA-EG4.bam histjl.png
# Reading BAM...done.
# Saving fig to histjl.png...done.
# qualhist.jl /Users/james/projects/hla/capture/bam38/PC-HLA-EG4.bam histjl.png  
# 29.42s user 1.53s system 99% cpu 31.158 total

# This implies:
# Warmup time 13.95 sec .
# Time to complete after warmup: 15.05 sec
